File: custom_components/kia_access/range.py
"""Drive-range reachability - Python port of core/range.js.

Kept line-for-line with the JS; both are exercised in CI
(test/range.test.js / test/range_test.py).
"""
from __future__ import annotations

import math

DEFAULTS = {
    "factor": 0.92,
    "reservePct": 10,
    "reserveKm": None,
    "roundTrip": False,
    "circleFactor": 0.85,
}


def _opt(o, k):
    o = o or {}
    v = o.get(k)
    return DEFAULTS[k] if v is None else v


def _num(v):
    try:
        f = float(v)
    except (TypeError, ValueError):
        return None
    return f if f == f and f not in (float("inf"), float("-inf")) else None


def reach(range_km, o=None):
    """Usable one-way (or half, round trip) drive distance in km, or None."""
    r = _num(range_km)
    if r is None or r <= 0:
        return None
    reserve_km = _opt(o, "reserveKm")
    if reserve_km is not None:
        usable = r - float(reserve_km)
    else:
        usable = r * (1 - float(_opt(o, "reservePct")) / 100)
    if not usable > 0:
        return 0
    d = usable * float(_opt(o, "factor"))
    return d / 2 if _opt(o, "roundTrip") else d


R_EARTH_KM = 6371.0088
D2R = math.pi / 180


def haversine_km(lat1, lon1, lat2, lon2):
    d_lat = (lat2 - lat1) * D2R
    d_lon = (lon2 - lon1) * D2R
    a = (
        math.sin(d_lat / 2) ** 2
        + math.cos(lat1 * D2R) * math.cos(lat2 * D2R) * math.sin(d_lon / 2) ** 2
    )
    return R_EARTH_KM * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))


def bearing_deg(lat1, lon1, lat2, lon2):
    y = math.sin((lon2 - lon1) * D2R) * math.cos(lat2 * D2R)
    x = math.cos(lat1 * D2R) * math.sin(lat2 * D2R) - math.sin(lat1 * D2R) * math.cos(
        lat2 * D2R
    ) * math.cos((lon2 - lon1) * D2R)
    return (math.atan2(y, x) / D2R + 360) % 360


def poi_status(lat, lon, pois, reach_km, trip=None):
    """Which saved places are in reach, nearest first.

    trip {batteryPct, rangeKm, roadFactor} also estimates arrival SoC.
    """
    if _num(lat) is None or _num(lon) is None or not isinstance(pois, list):
        return []
    trip = trip or {}
    pct = _num(trip.get("batteryPct"))
    rng = _num(trip.get("rangeKm"))
    road = _num(trip.get("roadFactor")) or 1.3
    can_arrive = pct is not None and rng is not None and rng > 0
    out = []
    for p in pois:
        if not p or _num(p.get("lat")) is None or _num(p.get("lon")) is None:
            continue
        km = haversine_km(lat, lon, float(p["lat"]), float(p["lon"]))
        arrival_pct = (
            round(max(0, pct * (1 - (km * road) / rng))) if can_arrive else None
        )
        out.append(
            {
                "name": p.get("name") or "",
                "km": km,
                "reachable": reach_km is not None and km <= reach_km,
                "marginKm": (reach_km - km) if reach_km is not None else None,
                "bearing": bearing_deg(lat, lon, float(p["lat"]), float(p["lon"])),
                "arrivalPct": arrival_pct,
            }
        )
    out.sort(key=lambda x: x["km"])
    return out


def circle_ring(lat, lon, km, n=64):
    """GeoJSON polygon ring (lon/lat) approximating a km-radius circle."""
    ring = []
    lat_r = km / 111.32
    lon_r = km / (111.32 * math.cos(lat * D2R) or 1e-6)
    for i in range(n + 1):
        t = (i / n) * 2 * math.pi
        ring.append([lon + lon_r * math.sin(t), lat + lat_r * math.cos(t)])
    return ring


def summary(car_lat, car_lon, range_km, pois=None, o=None):
    o = o or {}
    one = reach(range_km, {**o, "roundTrip": False})
    rnd = reach(range_km, {**o, "roundTrip": True})
    rt = bool(_opt(o, "roundTrip"))
    active = rnd if rt else one
    trip = {
        "batteryPct": o.get("batteryPct"),
        "rangeKm": range_km,
        "roadFactor": o.get("roadFactor"),
    }
    return {
        "roundTrip": rt,
        "oneWayKm": one,
        "roundTripKm": rnd,
        "reachKm": active,
        "pois": poi_status(car_lat, car_lon, pois or [], active, trip),
        "circle": circle_ring(car_lat, car_lon, active)
        if _num(car_lat) is not None and active
        else None,
    }

File: core/range.js
/* Drive-range reachability — shared by the MagicMirror module and the HA
 * integration. Turns the car's own range estimate into "how far can I get"
 * (one-way or there-and-back), tells you which saved places are in reach, and
 * builds a plain circle for when a road-network isochrone isn't available.
 *
 * Pure: no DOM, no `this`, no network. The isochrone fetch lives in the
 * callers (node_helper / coordinator); this module just does the maths.
 */
(function (root, factory) {
  if (typeof module === "object" && module.exports) module.exports = factory();
  else root.KiaAccessRange = factory();
})(typeof self !== "undefined" ? self : this, function () {
  "use strict";

  var DEFAULTS = {
    factor: 0.92,      // haircut on the car's guess-o-meter (cold air, a long pull)
    reservePct: 10,    // arrive with this much charge left…
    reserveKm: null,   // …or an absolute distance, if set (wins over reservePct)
    roundTrip: false,  // halve the distance for "there AND back"
    circleFactor: 0.85 // extra haircut for the straight-line circle fallback
  };

  function opt(o, k) {
    o = o || {};
    return o[k] === undefined || o[k] === null ? DEFAULTS[k] : o[k];
  }

  /** usable one-way (or half, round trip) drive distance in km, or null. */
  function reach(rangeKm, o) {
    var r = Number(rangeKm);
    if (!isFinite(r) || r <= 0) return null;
    var reserveKm = opt(o, "reserveKm");
    var usable =
      reserveKm != null
        ? r - Number(reserveKm)
        : r * (1 - Number(opt(o, "reservePct")) / 100);
    if (!(usable > 0)) return 0;
    var d = usable * Number(opt(o, "factor"));
    return opt(o, "roundTrip") ? d / 2 : d;
  }

  var R_EARTH_KM = 6371.0088;
  var D2R = Math.PI / 180;

  /** great-circle distance in km */
  function haversineKm(lat1, lon1, lat2, lon2) {
    var dLat = (lat2 - lat1) * D2R;
    var dLon = (lon2 - lon1) * D2R;
    var a =
      Math.sin(dLat / 2) * Math.sin(dLat / 2) +
      Math.cos(lat1 * D2R) * Math.cos(lat2 * D2R) *
        Math.sin(dLon / 2) * Math.sin(dLon / 2);
    return R_EARTH_KM * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  }

  /** initial bearing (degrees, 0 = N) from point 1 to point 2 */
  function bearingDeg(lat1, lon1, lat2, lon2) {
    var y = Math.sin((lon2 - lon1) * D2R) * Math.cos(lat2 * D2R);
    var x =
      Math.cos(lat1 * D2R) * Math.sin(lat2 * D2R) -
      Math.sin(lat1 * D2R) * Math.cos(lat2 * D2R) * Math.cos((lon2 - lon1) * D2R);
    return (Math.atan2(y, x) / D2R + 360) % 360;
  }

  /**
   * Which saved places are in reach.
   * @param {number} lat @param {number} lon  the car
   * @param {Array<{name,lat,lon}>} pois
   * @param {number} reachKm  from reach()
   * @param {object} [trip] { batteryPct, rangeKm, roadFactor } to also estimate
   *        the state-of-charge you'd arrive with
   * @returns sorted nearest-first:
   *   [{ name, km, reachable, marginKm, bearing, arrivalPct }]
   *   marginKm > 0 = spare, < 0 = short by that much; arrivalPct null if no trip
   */
  function poiStatus(lat, lon, pois, reachKm, trip) {
    if (!isFinite(lat) || !isFinite(lon) || !Array.isArray(pois)) return [];
    trip = trip || {};
    var pct = Number(trip.batteryPct);
    var rng = Number(trip.rangeKm);
    var road = Number(trip.roadFactor) || 1.3;
    var canArrive = isFinite(pct) && isFinite(rng) && rng > 0;
    var out = [];
    pois.forEach(function (p) {
      if (!p || p.lat == null || p.lon == null) return;
      var pla = Number(p.lat), plo = Number(p.lon);
      if (!isFinite(pla) || !isFinite(plo)) return;
      var km = haversineKm(lat, lon, pla, plo);
      var arrivalPct = null;
      if (canArrive) {
        arrivalPct = Math.round(Math.max(0, pct * (1 - (km * road) / rng)));
      }
      out.push({
        name: p.name || "",
        km: km,
        reachable: reachKm != null && km <= reachKm,
        marginKm: reachKm != null ? reachKm - km : null,
        bearing: bearingDeg(lat, lon, pla, plo),
        arrivalPct: arrivalPct
      });
    });
    out.sort(function (a, b) { return a.km - b.km; });
    return out;
  }

  /** GeoJSON polygon ring (lon/lat pairs) approximating a `km`-radius circle —
   *  the fallback when a road-network isochrone can't be had. */
  function circleRing(lat, lon, km, n) {
    n = n || 64;
    var ring = [];
    var latR = km / 111.32;
    var lonR = km / (111.32 * Math.cos(lat * D2R) || 1e-6);
    for (var i = 0; i <= n; i++) {
      var t = (i / n) * 2 * Math.PI;
      ring.push([lon + lonR * Math.sin(t), lat + latR * Math.cos(t)]);
    }
    return ring;
  }

  /** Everything a surface needs for one render. */
  function summary(carLat, carLon, rangeKm, pois, o) {
    o = o || {};
    var one = reach(rangeKm, Object.assign({}, o, { roundTrip: false }));
    var round = reach(rangeKm, Object.assign({}, o, { roundTrip: true }));
    var rt = opt(o, "roundTrip");
    var active = rt ? round : one;
    var trip = {
      batteryPct: o.batteryPct,
      rangeKm: rangeKm,
      roadFactor: o.roadFactor
    };
    return {
      roundTrip: !!rt,
      oneWayKm: one,
      roundTripKm: round,
      reachKm: active,
      pois: poiStatus(carLat, carLon, pois || [], active, trip),
      circle: (isFinite(carLat) && active) ? circleRing(carLat, carLon, active) : null
    };
  }

  return {
    DEFAULTS: DEFAULTS,
    reach: reach,
    haversineKm: haversineKm,
    bearingDeg: bearingDeg,
    poiStatus: poiStatus,
    circleRing: circleRing,
    summary: summary
  };
});
